Add process time counters

sywhang opened the pull request:
For #20372. 

This adds a Runtime counter for process time (user time + kernel/system time). 

cc @jkotas

jkotas commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.OSX.cs:
```diff
@@ -0,0 +1,79 @@
+
+        // Constant from sys\resource.h
+        private const int RUSAGE_INFO_V3 = 3;
+        private const string OSX_LIBPROC = "libproc";
```
We do not want to PInvoke the Unix APIs directly. They have unstable ABI.

We always PInvoke thin shims over the Unix ABI in `System.Native*` libraries. More details: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md#unix-shims

I think you should be able to use the existing `SystemNative_GetCpuUtilization` shim for this: https://github.com/dotnet/corefx/search?q=SystemNative_GetCpuUtilization&unscoped_q=SystemNative_GetCpuUtilization . Note that it is fine to use it in CoreLib even though it is implemented in CoreFX.

sywhang replied:
Thanks for the suggestion Jan! And sorry for the delayed response :-( I was looking at adding some other APIs and came back to this recently. 

Quick question about that surface though -- It seems to me that this surface seems to be calculating "CPU usage in percentage". In particular, these lines (https://github.com/dotnet/corefx/blob/c1328f49f69d501618f1d867265efacedf107f23/src/Native/Unix/System.Native/pal_time.c#L194):
```
    if (cpuTotalTime > 0 && cpuBusyTime > 0)
    {
        cpuUtilization = (int32_t)(cpuBusyTime / cpuTotalTime);
    }
```
seems like it's calculating the ratio of busy time vs total time instead of the percentage. And since it's casted to an int, it always returns 0 when I tested it. 

Am I misinterpreting the intent of this surface? Or is this an actual issue?

jkotas replied:
You can add a new API to the shim; or fixup the existing one to be more general - whatever works better.

My key point is that you should be able to just call the API, without parsing a bunch of stuff.

sywhang replied:
Got it. Thanks!

jkotas commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.cs:
```diff
@@ -0,0 +1,24 @@
+	{
+		internal static long GetProcessTimes()
+		{
+			long _creation;
```
Local variables should not start with `_`.

jkotas commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.cs:
```diff
@@ -0,0 +1,24 @@
+			long _user;
+			long _kernel;
+			
+			Interop.Kernel32.GetProcessTimes(Interop.Kernel32.GetCurrentProcess(), out _creation, out _exit, out _user, out _kernel);
```
You can use `_` placeholder for the values you do not care about.

jkotas commented on src/System.Private.CoreLib/src/System/Environment.CoreCLR.cs:
```diff
@@ -88,6 +88,9 @@ public static string[] GetCommandLineArgs()
         private static extern int GetProcessorCount();
 
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        internal static extern int GetCurrentProcessId();
```
This should not be needed once you incorporate the other feedback.

AaronRobinsonMSFT commented on src/System.Private.CoreLib/shared/Interop/Windows/Kernel32/Interop.GetProcessTimes.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    internal partial class Kernel32
+    {
+        [DllImport(Libraries.Kernel32, CharSet = CharSet.Unicode, SetLastError = true)]
```
The `CharSet` property isn't needed.

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.OSX.cs:
```diff
@@ -0,0 +1,79 @@
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct rusage_info_v3
+        {
+            internal fixed byte     ri_uuid[16];
```
The struct is `private` inside of an `internal` class. The fields can be marked `public`.

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.OSX.cs:
```diff
@@ -0,0 +1,79 @@
+        /// <param name="buffer">A buffer to be filled with rusage_info data</param>
+        /// <returns>Returns 0 on success; on fail, -1 and errno is set with the error code</returns>
+        [DllImport(OSX_LIBPROC, SetLastError = true)]
+        private static extern unsafe int proc_pid_rusage(
```
This should go away based on @jkotas's suggestion, but for future reference the above name can made to follow the coreclr style by using the `EntryPoint` property on the `DllImport` attribute.

jorive commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.Linux.cs:
```diff
@@ -0,0 +1,72 @@
+{
+	internal sealed class RuntimeEventSourceHelper
+	{
+		internal static long GetProcessTimes()
```
nit: The formatting of this function seems odd.

stephentoub replied:
Yes, it looks like spaces have been intermixed with tabs.  All tabs should be converted to spaces.

stephentoub commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSource.cs:
```diff
@@ -50,6 +51,7 @@ protected override void OnEventCommand(System.Diagnostics.Tracing.EventCommandEv
                     // overhead by at all times even when counters aren't enabled.
                     _counters = new EventCounter[] {
                         // TODO: process info counters
```
Can this TODO be removed, or are there more to be added after this one?  Is there an issue number to associate with this?

stephentoub commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.Linux.cs:
```diff
@@ -0,0 +1,72 @@
+		internal static long GetProcessTimes()
+		{
+            long pid = Environment.GetCurrentProcessId();
+            using (FileStream fs = new FileStream($"/proc/{pid}/stat", FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 1))
```
We have helpers that do all this in corefx, e.g.
https://github.com/dotnet/corefx/blob/b8bc4ff80c5f7baa681e8a569d367356957ba78a/src/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.cs#L195
Should we move that to the shared corelib folder and reuse those routines?

jkotas replied:
None of this should be needed once the rest of the feedback is addresses.

stephentoub replied:
> None of this should be needed once the rest of the feedback is addresses.

Ah, great.  I hadn't read the other feedback yet.  I'll just stop commenting on this until it's addressed :)

stephentoub commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.Linux.cs:
```diff
@@ -0,0 +1,72 @@
+                byte[] text = ReadAllBytesUnknownLength(fs);
+                string s = System.Text.Encoding.UTF8.GetString(text, 0, text.Length);
+                long _user = Convert.ToInt64(s.Split(" ")[13]);
+                long _system = Convert.ToInt64(s.Split(" ")[14]);
```
Split is relatively expensive.

jorive replied:
@stephentoub What would you recommend? What would be a performant option here? Using a regex?

jkotas replied:
None of this should be needed once https://github.com/dotnet/coreclr/pull/22901#discussion_r261051884 is addressed.

stephentoub commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.Linux.cs:
```diff
@@ -0,0 +1,72 @@
+                        uint newLength = (uint)buffer.Length * 2;
+                        if (newLength > MaxByteArrayLength)
+                        {
+                            newLength = (uint)Math.Max(MaxByteArrayLength, buffer.Length + 1);
```
If the buffer ever got to be > than 1 or 2MBs, whatever ArrayPool's current max is, then this is going to result in allocating a new large array for every additional byte.

stephentoub commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.Linux.cs:
```diff
@@ -0,0 +1,72 @@
+                        if (rentedArray != null)
+                        {
+                            ArrayPool<byte>.Shared.Return(rentedArray);
+                        }
```
If some exception were to occur here, after having returned the array and before reassigning rentedArray, the finally block will end up returning the same buffer again.  Given that thread aborts don't exist, it's not clear there's any such exception that could happen, but I'd sleep better if this were instead:
```C#
byte[] tmp = ArrayPool<byte>.Shared.Rent((int)newLength);
buffer.CopyTo(tmp);
byte[] oldArray = rentedArray;
buffer = rentedArray = tmp;
if (oldArray != null) ArrayPool<byte>.Shared.Return(oldArray);
```

sywhang commented:
Thanks everyone for the feedback! I'm closing this PR as the EventCounter APIs diverged a lot since this and I worked off a separate branch (it should've addressed most of the concerns raised in this PR if not all) and opened #23680 instead.
